include/xtensor/xfunction.hpp

```diff
@@ -199,7 +235,8 @@ namespace xt
      * @tparam CT the closure types for arguments of the function
      */
     template <class F, class R, class... CT>
-    class xfunction_base : private xconst_iterable<xfunction_base<F, R, CT...>>
+    class xfunction_base : private xconst_iterable<xfunction_base<F, R, CT...>>,
+                           private xfunction_cache<detail::promote_index<typename std::decay_t<CT>::shape_type...>>
```

Comment:
Composition (i.e. having a private `xfunction_cache`  member) would be better, directly accessing and changing data members of base class from ionheriting classes may be confusing and exposes base class implementation details to the inheriting one. Sorry I missed that during the first review!